packages/canvas-runtime/src/hooks/useHintOverlays.tsx

```diff
@@ -36,51 +47,67 @@ export function removeHintOverlays(overlayIds: string[]) {
+        paddingBottom,
+      } = getComputedStyle(comp);
+      const box = props.box({
+        dimension: {
+          height: compCoords.height,
+          width: compCoords.width,
+          marginBottom: parseFloat(marginBottom),
+          marginLeft: parseFloat(marginLeft),
+          marginRight: parseFloat(marginRight),
+          marginTop: parseFloat(marginTop),
+          paddingBottom: parseFloat(paddingBottom),
+          paddingTop: parseFloat(paddingTop),
+          paddingLeft: parseFloat(paddingLeft),
+          paddingRight: parseFloat(paddingRight),
+        },
+      });
+      return { box, bodyCoords, compCoords };
+    }
+  }
+  return null;
+}
+
 const HintOverlayBox: React.FC<HintOverlay & { scale: number }> = (props) => {
-  const [box, setBox] = useState<ReturnType<HintOverlay["box"]> | null>(null);
-  const [bodyPosition, setBodyPosition] = useState<Position | null>(null);
-  const [compPosition, setCompPosition] = useState<Position | null>(null);
+  const boxDimensions = calculateBoxDimensions(props);
+  let initialBox = null, initialBody = null, initialComp = null;
+  if (boxDimensions) {
+    initialBox = boxDimensions.box;
+    initialBody = boxDimensions.bodyCoords;
+    initialComp = boxDimensions.compCoords;
+  }
+  const [box, setBox] = useState<ReturnType<HintOverlay["box"]> | null>(initialBox);
+  const [bodyPosition, setBodyPosition] = useState<Position | null>(initialBody);
+  const [compPosition, setCompPosition] = useState<Position | null>(initialComp);
   useEffect(() => {
-    if (canvasComponentStore[props.compId]) {
-      if (
-        canvasComponentStore["body"].ref.current &&
-        canvasComponentStore[props.compId].ref.current
-      ) {
-        const body = canvasComponentStore["body"].ref.current;
-        const comp = canvasComponentStore[props.compId].ref.current!;
-        const bodyCoords = getCoords(body);
-        const compCoords = getCoords(comp);
-        const {
-          marginBottom,
-          marginTop,
-          marginLeft,
-          marginRight,
-          paddingLeft,
-          paddingRight,
-          paddingTop,
-          paddingBottom,
-        } = getComputedStyle(comp);
-        const box = props.box({
-          dimension: {
-            height: compCoords.height,
-            width: compCoords.width,
-            marginBottom: parseFloat(marginBottom),
-            marginLeft: parseFloat(marginLeft),
-            marginRight: parseFloat(marginRight),
-            marginTop: parseFloat(marginTop),
-            paddingBottom: parseFloat(paddingBottom),
-            paddingTop: parseFloat(paddingTop),
-            paddingLeft: parseFloat(paddingLeft),
-            paddingRight: parseFloat(paddingRight),
-          },
-        });
-        setBox(box);
-        setBodyPosition({ top: bodyCoords.top, left: bodyCoords.left });
-        setCompPosition({ top: compCoords.top, left: compCoords.left });
-      }
+    const boxDimensions = calculateBoxDimensions(props);
```

Comment:
After this refactoring, it's evident that we can make this code even better, by moving all the calculations in the `useMemo` hook. If we use `useMemo`, we won't need `useState` & `useEffect` hook. The psuedo-code for the change is below.

```jsx
// all other hooks removed
// only one useMemo
const {box, bodyPosition, compPosition} = useMemo(()=>{ return calculateBoxDimensions(props) }, [props]);
return <></>
```

Reply:
fixed